### DISCUSSION TITLE
Ideas

### DISCUSSION THREAD

**Discussion by SuperLlama88888** [SuperLlama88888/holoprint#7] [General]
### To-Do, planned features
- Could we use #define USE_COLOR_MASK in a material to apply opacity?!?!? Except it seems to be CHANGE_COLOR instead and COLOR_SECOND_TEXTURE in horses
- Putting the pack creator inside an addon and creating it inside the addon? But saving the file won't work... unless we use @minecraft/server-net to upload it to a server but that's only BDS (I think)
  - Since HoloPrint is made in JavaScript it'd mostly be portable inside an addon, but web APIs would need replacements with other libraries. I also hope we can access vanilla textures in addons.
  - I haven't looked at the capabilities of the Scripting API with the Editor but if we could put it in there, it would be even better!
  - The docs mention the "experimental_custom_ui" capability in manifest.json that apparently allows HTML to make UIs - I'm thinking it's OreUI but I haven't seen anyone talk about it so I assume it must nonfunctional. If anyone knows anything about this please reach out.
  - If Jigarbov Production's special "registry" between addons is made publicly available, it could allow another addon on the target world to load the data and somehow render the ghost blocks? IDK how we'd render it then because currently the hologram is hard-baked into the geometry file, and IDK if using hundreds/thousands of entities each rendering a single block would be performant.
  - Even if this registry is also accessible in resource packs, constructing the geometry on the fly would be hard.
- ~~Changing armour stand item to show different models~~
- ~~Hologram tint overlay toggleable~~
- ~~Item frame/something other than armour stand??? Item frame is impossible in a resource pack alone, other entities would be possible but they would move. Tamed wolves perhaps?~~
- Using block models from JE?
  - This would require Bedrock-Java mappings, I think Amulet's super-format could be used. The Amulet format is a universal format for all MC versions in both editions in the last 10 years so would work well.
    - Again, the problem with autogeneration of block geometries is that different textures from resource packs and custom blocks from addons would still need to be done the classic way - how we're doing it currently.
    - Also, due to it being a webpage we want to really minimise performance. We don't want to crash peoples' phones when they try to generate a pack. I fear that using these existing, big-data, high-level solutions could make it less accessible to some players.
    - Note that in Java Edition, block model geometries are available in resource packs already.
      - I've been using these geometries as a guide, and I think if we can achieve what we need just with this repo it will be faster than relying on other projects (which would be unspecialised)...
- Full auto-generation of block geometries (including UVs) all within Bedrock Edition???
  - Use the scripting API to generate all blocks in the world with all states - much like the Java Edition debug world
    - Problem: Fences, glass panes, iron bars, and stairs don't rely on block states for their visual appearance, so it'll have to place blocks around them then
  - Using the structure block .glb export mode only on Windows 10, we export all the blocks into JSON (glb can be turned easily into gltf which is json)
  - Then, when a structure file is uploaded, it looks up the block geometry from the previous step.
  - Many features of HoloPrint such as block scaling, adding the outlines, creating a minified texture atlas, armour stand + player controls etc. would still be needed.
  - Support for resource packs and custom blocks from addons would have to be done in a different way.
  - Could this be done in GitHub Actions on a Windows machine???
    - The two biggest problems are getting Minecraft on a GH Actions machine and interacting with the client without a screen.
    - Looking at the (awful) protocol docs for the [`StructureTemplateDataResponsePacket`](https://mojang.github.io/bedrock-protocol-docs/html/StructureTemplateDataResponsePacket.html) seems like only the structure NBT is transmitted, not the .glb.
    - This is done entirely on the client, which would be harder to simulate in GitHub Actions. Perhaps something fun with Amethyst?
    - Further investigation (i.e. decompilation, static analysis) would be required if we wanted to go down this path, but I think it'd be the most rewarding solution in the long run.
    - Looks like it's a class MinecraftglTFExporter stored in D:\a\_work\1\s\handheld\src-client\common\3DExport\MinecraftglTFExporter.cpp
- ~~Hitting armour stands to change models? inspired by Indyficcient~~
  - ~~Since items are already used for other hologram controls, this makes it a bit simpler I think. may be annoying though? look at https://wiki.bedrock.dev/entities/non-mob-runtime-identifiers.html#minecraft-armor-stand~~
- ~~Fix hologram in smithing table UI~~
- Automated 3rd-party server resource pack patching? Modifying `%TEMP%\Minecraft Bedrock\minecraftpe\packcache\resource`? Would this work?!?!?

### About addons
- Behaviour packs would open so many possibilities for HoloPrint packs, but would also be inaccessible to people without access to the server's settings. I don't know much about behaviour packs but I'm guessing it would be quite powerful. Here's what I think we could do:
  - ~~Item frame as the base entity? Not possible in resource packs, and if we're using a behaviour pack we might as well provide an easier way to show holograms than using item frames.~~
  - ~~Cycling through layers with UI? Impossible in a resource pack~~
  - Pick-blocking ghost blocks: Maybe in an addon? We'll need to be able to detect middle clicks, do a raycast, then change what the player's holding.
- Regarding a generalised behaviour pack that interacts with generated resource packs: Pretty sure this would be possible. The server would just have to tell the armour stand to do something, but I don't have any experience with the Scripting API.

**Comment by KittyShizz** [comment]
Would love a raw command to hologram as so many sites like https://minecraftart.netlify.app/ can output builds as just commands
if thats possible? This might be one of the coolest things ive seen

Example File:
setblock ~11 ~44 ~22 coal_block 0  
fill ~11 ~44 ~23 ~11 ~44 ~24 concrete 15  
fill ~11 ~44 ~60 ~12 ~44 ~60 concrete_powder 15  
fill ~11 ~44 ~61 ~11 ~45 ~61 concrete 15  
fill ~11 ~44 ~62 ~14 ~44 ~62 concrete_powder 15 
setblock ~11 ~44 ~63 obsidian 0 
setblock ~11 ~44 ~64 concrete_powder 15  
fill ~11 ~44 ~68 ~11 ~44 ~69 concrete 15  
fill ~11 ~44 ~70 ~11 ~45 ~70 obsidian 0

**Reply by SuperLlama88888** [reply]
Thanks for the suggestion! Since the mentioned website is for Java Edition, I'd take a look at this feature while implementing #46, since I'll need a converter from Java -> Bedrock blocks. It won't be coming soon, sorry...

**Comment by Erudius** [comment]
Just wanted to drop another suggestion.  I'm loving the malleability, but is HoloPrint always on by default?
Right now, all the armor stands I have around the base for decoration/diorama are generating placement holos, and there's no way to use them as a regular armor stand without giving them all a brick. Reading the instructions, it doesn't seem there is other than putting a brick in the hand.

so... the suggestion by way of example:  Foxy uses the armor stand as a chunk tool by giving it a Banner.  Perhaps the armor stand could simply be an armor stand unless you put a certain piece of gear/armor on? (does color matter? could expand options for on-switch)

Thanks so much for what you've accomplished already!

***  UPDATE  ***
Tried the brick fix but it reloads as soon as you get out of range. Causes a fair bit of lag, so I've had to remove it from the server until I can get it resolved so that armor stands can be armor stands by default.

![Current Game 2025-01-21_16-43-45](https://github.com/user-attachments/assets/933aa30a-7569-43bf-834c-8cd38ec10ca5)
![Current Game 2025-01-21_16-50-18](https://github.com/user-attachments/assets/5c81dd7b-8b8c-4720-82e6-3b41bc988339)
![Current Game 2025-01-21_16-58-26](https://github.com/user-attachments/assets/bd67fbed-bcb8-4c8d-a5fa-12c0d9f1c57d)

**Reply by SuperLlama88888** [reply]
Hi, in the pack settings there's an option called "Punch to activate" which makes holograms be hidden until you punch the armour stand. I'll add info on that to the wiki now. Hope this helps!

**Reply by Erudius** [reply]
Thanks!  Yes, I totally missed that.  I'm wondering, would it be prudent to have that enabled by default?

*update*
It appears it also turns on merely by interacting with it so by even changing the pose, it turns it on, so it seems any new stands added while construction is taking place will display the holo also even if this switch is on.

**Reply by SuperLlama88888** [reply]
This is intended behaviour, even though its called "Punch to activate" it also allows changing the pose to activate. This is to expand options for players who can't punch armour stands for whatever reason.

Could you please elaborate on your last sentence? Newly-placed armour stands shouldn't show the hologram until interacted with in this mode. If they do, that's a bug (which I'm unable to reproduce).

**Reply by Erudius** [reply]
Basically... I tried it in a solo world.  It works great, punch to activate etc.  It's on the realm that this becomes an issue.

I've enabled punch to activate, but the holograms still come up on loading into the realm as well as when they are newly placed.  Of course I'm just 1 guy.  I couldn't tell you if it's a problem on other realms.  If you want I'll run a capture of what I'm doing.

**Reply by SuperLlama88888** [reply]
That's definitely a bug. Could you please create an issue and attach the pack you're using? A screen capture would also be great.

**Reply by Erudius** [reply]
Will do.  
It might be how Microsoft's Bedrock servers handle the .mcpack also.  From what I've been able to gather, they will accept  the upload of .mcpacks, but they will not allow global resource packs.
I'll put it all in the bug report.

**Reply by Erudius** [reply]
I did some research on my end. It's a conflict with Foxy's marker pack... what i use to find chunk borders.  If Foxy's has priority the chunk borders don't work. If Holoprint has priority, foxy's turns off, so I won't be able to use that chunk border tool.  I'll just need to switch between mods.  No worries.

... in which case ... maybe you could make a chunk border addition item to use with holo?  Some builds depend on redstone being in the same chunk.
the other thing I use Foxy's for is the Mob spawn spheres... if that's something you could add would be cool too.

**Reply by SuperLlama88888** [reply]
I'll think about adding a chunk display as well.

**Comment by Pete9xi** [comment]
It would be useful to set the offset at the moment there is no ability todo that so you either sink the amour stand down one layer if you have captured the terrian around a build(i tend to do the terrain as well as the build)

**Reply by SuperLlama88888** [reply]
Currently, you can [move the hologram](https://holoprint-mc.github.io/wiki/hologram-controls#move-hologram), but I will in future allow an offset to be specified during pack creation.

**Reply by SuperLlama88888** [reply]
Hi @Pete9xi, sorry it's taken so long, but you can now specify the initial offset during pack creation:
![image](https://github.com/user-attachments/assets/e3c7a33b-c80f-47b7-a262-abdcb33725bb)

**Reply by Pete9xi** [reply]
Fantastic that's going to be a useful feature

**Comment by LukeyMan2007** [comment]
Hi! I've been using Holoprint for a while and I've been loving it. This may seem kinda weird but would there be a way to change the hotkey to view the material list? I've also been using ty el's ui but that also has a hotkey with tab that cannot be changed. If there's no way to change the material list hotkey to anything else as an option I totally get it, thanks!

**Reply by SuperLlama88888** [reply]
Hmm, it's a bit difficult since there's no way to fully control the key used. We do have a few options though.
If you open up the resource pack and go to `ui/hud_screen.json`, you want to find "button.scoreboard" and replace that with "button.hide_tooltips". This should make the material list show when you press F10.

If you send your HoloPrint pack I'll be happy to change it for you.

**Reply by LukeyMan2007** [reply]
I'm sure I can figure it out, thank you so much you have no idea how helpful you are for doing all of this for the community it means a lot. Have a great night and thank you!

**Comment by Khb17** [comment]
Pro add the easy place mod is so cooooooool

**Comment by KWiK87** [comment]
This is a very cool tool. It is nice that you can define an initial offset.

It is also very nice that you can use multiple structures in the same pack

However, if you are working with multiple structures in the same pack, then it would be nice to be able to define a different initial offset for each structure.

**Reply by SuperLlama88888** [reply]
I like the idea, but making the UI work well will require a few changes. Thanks for the suggestion!

**Reply by SuperLlama88888** [reply]
Sorry @KWiK87, but I've decided against implementing this change. My reasoning is that it would reset the offset every time the structure is changed. I want changes you've made to the hologram to persist when changing structure, and making it reset the offset every time you change structure would go against this pattern.
For a temporary workaround, try remaking your structure file, but having lots of air blocks around the edges to add some padding.

**Comment by what-dude** [comment]
There are a lot of schematic builds out there, and I have been searching for, like hours, for a tool that can convert from schematic to .mcstructure, since holoprint only supports worlds, packs, or .mcstructres, but i haven't found anything that can convert that. Can you add a feature so holoprint can support multiple formats? Like .schem, .schematic, .litematic, .nbt, etc?

This would be sooooooooooooo useful.

**Reply by SuperLlama88888** [reply]
It'd planned for the future, but it's pretty hard since Minecraft Java and Bedrock have differences in how blocks and block states are named. For now I'd recommend putting your schematic inside a world, using [Chunker](https://www.chunker.app/) to convert the world to Bedrock Edition, then export your structure block as normal from there.

**Reply by what-dude** [reply]
Thanks for your response.
Let me know when that gets added to holoprint.

**Reply by raducal** [reply]
You should try out https://bloxelizer.com/. It can convert from .schem, .schematic etc to .mcstructure and works well with Holoprint

**Reply by SuperLlama88888** [reply]
I have tried it out before, and it works well. However because it's not open-source it would be difficult to add the structure conversion in the same way it does.

**Comment by Hazem4769** [comment]
Is it possible to add auto block placer like Java?

**Reply by SuperLlama88888** [reply]
Not in a resource pack, but I'm going to start work on a behaviour pack soon, which will have something like easyplace like it is in [Construct](https://github.com/ForestOfLight/Construct).

**Reply by Hazem4769** [reply]
This would be an amazing and truly useful update!
I hope your behavior pack won’t disable achievements like most addons do these days — that would make it even better!
Best of luck with your project! 😄
